core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/browsebars/BrowseBarComplete.java

```diff
@@ -85,141 +86,167 @@ private String printBreadCrumb() {
+    appendExtraInformation(breadcrumb);
+
+    result.append(breadcrumb);
+
+    result.append("</div>");
+
+    return result.toString();
+  }
+
+  private String getUserLanguage() {
+    if (userLanguage == null) {
+      userLanguage = (getMainSessionController() == null) ? "" : getMainSessionController()
+          .getFavoriteLanguage();
+    }
+    return userLanguage;
+  }
+
+  private String getSpacePath(SilverpeasComponentInstance componentInst) {
+    List<SpaceInstLight> spaces = Collections.emptyList();
+    OrganizationController organizationController =
+        OrganizationControllerProvider.getOrganisationController();
+    if (componentInst != null && !componentInst.isPersonal()) {
+      spaces = organizationController.getPathToComponent(getComponentId());
+    } else if (componentInst == null) {
+      spaces = organizationController.getPathToSpace(getSpaceId());
+    }
+    StringBuilder breadcrumb = new StringBuilder();
+    for (SpaceInstLight spaceInst : spaces) {
+      String spaceId = spaceInst.getId();
+      String href = "javascript:" + getSpaceJavascriptCallback() + "('" + spaceId + "')";
+      if (!isClickable()) {
+        href = "#";
+      }
+
+      if (breadcrumb.length() > 0) {
+        breadcrumb.append(CONNECTOR);
+      }
+      breadcrumb.append("<a href=\"").append(href).append("\"");
+      breadcrumb.append(" class=\"space\"");
+      breadcrumb.append(" id=\"space").append(spaceId).append("\"");
+      breadcrumb.append(">");
+      breadcrumb.append(Encode.forHtml(spaceInst.getName(getUserLanguage())));
+      breadcrumb.append("</a>");
+    }
+    return breadcrumb.toString();
+  }
+
+  private String getComponent(SilverpeasComponentInstance componentInst) {
+    StringBuilder breadcrumb = new StringBuilder();
+    if (componentInst != null) {
+      // Display component's label
+      if (!componentInst.isPersonal()) {
+        breadcrumb.append(CONNECTOR);
+      }
+      breadcrumb.append("<a href=\"");
+      if (!isClickable()) {
+        breadcrumb.append("#");
+      } else if (StringUtil.isDefined(getComponentJavascriptCallback())) {
+        breadcrumb.append("javascript:").append(getComponentJavascriptCallback()).append("('")
+            .append(getComponentId()).append("')");
+      } else {
+        breadcrumb.append(URLUtil.getApplicationURL()).append(URLUtil.getURL(getSpaceId(),
+            getComponentId()));
+        if (ignoreComponentLink()) {
+          breadcrumb.append("Main");
+        } else {
+          breadcrumb.append(getComponentLink());
+        }
+      }
+      breadcrumb.append("\"");
+      breadcrumb.append(" class=\"component\"");
+      breadcrumb.append(" id=\"bc_").append(componentInst.getId()).append("\"");
+      breadcrumb.append(">");
+      breadcrumb.append(Encode.forHtml(componentInst.getLabel(getUserLanguage())));
+      breadcrumb.append("</a>");
+    }
+    return breadcrumb.toString();
+  }
+
+  private void appendPath(StringBuilder breadcrumb) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 20 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS3776)